src/morphology/english/determineStem.js

```diff
@@ -0,0 +1,154 @@
+		if ( endsWithIng( baseIfPluralNoun ) ) {
+			return buildOneFormFromRegex( baseIfPluralNoun, createRulesFromMorphologyData( regexVerb.ingFormToInfinitive ) );
+		}
+		return baseIfPluralNoun;
+	}
+
+	// Check if the word ends with "ic", "ical" or "ically". Return the "ical" form for consistency.
+	const regexAdjective = morphologyData.adjectives.regexAdjective;
+	const baseIfIcally = buildOneFormFromRegex( word, createRulesFromMorphologyData( regexAdjective.icallyToBase ) );
+	if ( ! isUndefined( baseIfIcally ) ) {
+		return baseIfIcally;
+	}
+
+	// No more quick guesses, we have to determine a verbal infinitive and an adjectival base.
+	const possibleRegularBases = [];
+
+	// Verbal infinitive.
+	const sFormToInfinitiveRegex = createRulesFromMorphologyData( regexVerb.sFormToInfinitive );
+	const ingFormToInfinitiveRegex = createRulesFromMorphologyData( regexVerb.ingFormToInfinitive );
+	const edFormToInfinitiveRegex = createRulesFromMorphologyData( regexVerb.edFormToInfinitive );
+
+	const baseIfVerb = getInfinitive( word, sFormToInfinitiveRegex, ingFormToInfinitiveRegex, edFormToInfinitiveRegex ).infinitive;
+
+	possibleRegularBases.push( baseIfVerb );
+
+	// Adjectival base.
+	const comparativeToBaseRegex = createRulesFromMorphologyData( regexAdjective.comparativeToBase );
+	const superlativeToBaseRegex = createRulesFromMorphologyData( regexAdjective.superlativeToBase );
+	const adverbToBaseRegex = createRulesFromMorphologyData( regexAdjective.adverbToBase );
+	const lyExceptions = morphologyData.adjectives.stopAdverbs;
+
+	const baseIfAdjective = getBase( word, comparativeToBaseRegex, superlativeToBaseRegex, adverbToBaseRegex, lyExceptions ).base;
+	possibleRegularBases.push( baseIfAdjective );
+
+	return findShortestAndAlphabeticallyFirst( possibleRegularBases );
+}
+
+/**
+ * Returns the stem of the input word using the morphologyData (language-specific).
+ *
+ * @param   {string} word           The word to get the stem for.
+ * @param   {Object} morphologyData The available morphology data per language (false if unavailable).
+ *
+ * @returns {string} Stemmed (or base) form of the word.
+ */
+export function determineStem( word, morphologyData ) {
+	if ( ! morphologyData ) {
+		return word;
+	}
+
+	let bestBase = word;
+
+	const nounMorphology = morphologyData.nouns;
+	let verbMorphology = morphologyData.verbs;
+
+	let irregulars = [].concat(
+		nounMorphology.irregularNouns,
+		morphologyData.adjectives.irregularAdjectives,
+	);
+
+	const baseIfPossessive = buildOneFormFromRegex( word, createRulesFromMorphologyData( nounMorphology.regexNoun.possessiveToBase ) );
+	// If the word is a possessive, it can only be a noun or an ing-noun.
+	if ( ! isUndefined( baseIfPossessive ) ) {
+		bestBase = baseIfPossessive;
+
+		// Only check for nouns.
+		irregulars = nounMorphology.irregularNouns;
+		verbMorphology = false;
+	}
+
+	const baseIfIrregular = determineIrregularStem( bestBase, irregulars, verbMorphology );
```

Comment:
If you split the `determineIrregularStem` function, I would change the function like this:
( I may got carried away 😅 ).
```js
if ( ! morphology ) {
     return word;
}

const nounIrregulars = morphologyData.nouns.irregularNouns;
const adjectiveIrregulars = morphologyData.adjectives.irregularAdjectives;

const baseIfPossessive = buildOneFormFromRegex( word, createRulesFromMorphologyData( nounMorphology.regexNoun.possessiveToBase ) );

let stem;

// Determine if this is an irregular noun, adjective or verb. 
if ( ! isUndefined( baseIfPossessive ) ) {
    // The word is a possessive, it can only be a noun or an ing-noun; 
    stem = determineIrregularStem( baseIfPossessive, nounIrregulars );
} else {
    // Word can be a noun, adjective or verb.
    stem = determineIrregularStem( word, nounIrregulars ) || 
           determineIrregularStem( word, adjectiveIrregulars ) || 
           determineIrregularVerbStem( word, verbMorphology );
}
if ( stem ) {
    // Found the stem of an irregular word! Return it.
    return stem;
}

// Treat the word as a regular noun, adjective or verb.

if ( ! isUndefined( baseIfPossessive ) ) {
     // Possessive: treat it as a regular noun or ing-noun.
     return determineRegularStem( baseIfPossessive, morphologyData );
} else {
     return determineRegularStem( word, morphologyData ); 
}
```

Reply:
Nice! 👍